### PR TITLE
[Routing] Add support for aliasing routes

### LINKED ISSUE #6088: [Routing] Add support for alias routes
``` yaml
# routes.yml

app_index:
    pattern: /

vendor_bundle_index:
    alias: app_index
```

### PR DESCRIPTION
**Pull request by Lctrs**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no 
| Tickets       | Fix #6088
| License       | MIT
| Doc PR        | None yet

### CODE REVIEW

**Comment by Lctrs** [comment]
Will update changelog later. Created it first to gather reviews.

File: src/Symfony/Component/Routing/Alias.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    {
+        $this->name = $name;
+    }
+
+    /**
+     * Returns the name of this alias.
+     *
+     * @return string The alias name
+     */
+    public function __toString()
```
Comment (PierreRambaud) [inline src/Symfony/Component/Routing/Alias.php:31]:
```suggestion
    public function __toString(): string
```

File: src/Symfony/Component/Routing/Loader/YamlFileLoader.php [hunk 2]
```diff
@@ -243,6 +249,12 @@ protected function validate($config, string $name, string $path)
         if ($extraKeys = array_diff(array_keys($config), self::$availableKeys)) {
             throw new \InvalidArgumentException(sprintf('The routing file "%s" contains unsupported keys for "%s": "%s". Expected one of: "%s".', $path, $name, implode('", "', $extraKeys), implode('", "', self::$availableKeys)));
         }
+        if (isset($config['alias'])) {
+            if (1 === \count($config)) {
+                return;
```
Comment (PierreRambaud) [inline src/Symfony/Component/Routing/Loader/YamlFileLoader.php:254]:
```suggestion
                // No need to continue, if an alias is set, it's the only desired key
                return;
```

Add a comment explaining why there is a return here :)

File: src/Symfony/Component/Routing/RouteCollection.php [hunk 3]
```diff
@@ -303,4 +322,65 @@ public function addResource(ResourceInterface $resource)
...
+            $route = new Alias($route);
+        } elseif (!$route instanceof Alias) {
+            throw new InvalidArgumentException('$route must be a string, or an Alias object.');
+        }
+
+        if ($alias === (string) $route) {
+            throw new InvalidArgumentException(sprintf('An alias can not reference itself, got a circular reference on "%s".', $alias));
+        }
+
+        unset($this->routes[$alias], $this->priorities[$alias]);
```
Comment (PierreRambaud) [inline src/Symfony/Component/Routing/RouteCollection.php:347]: Add a comment to explain why you unset these keys in these array?

File: src/Symfony/Component/Routing/RouteCollection.php [hunk 4]
```diff
@@ -303,4 +322,65 @@ public function addResource(ResourceInterface $resource)
...
+                throw new RouteNotFoundException(sprintf('Target route "%s" for alias "%s" is not defined.', $target, $alias));
+            }
+
+            $resolvedAliases[$alias] = $resolved;
+        }
+
+        return $resolvedAliases;
+    }
+
+    private function resolveAlias(string $alias): ?string
```
Comment (PierreRambaud) [inline src/Symfony/Component/Routing/RouteCollection.php:378]:
```suggestion

    /**
     * @return string|null 
     */
    private function resolveAlias(string $alias): ?string
```

**Comment by stof** [comment]
based on the discussion happening in https://github.com/symfony/symfony/issues/6088#issuecomment-484159219 and following comments, we should be able to mark aliases as deprecated, so that the UrlGenerator triggers a deprecation when trying to generate the URL for such alias (pointing to the new name).
This would have an impact on the way the feature is implemented (currently, you convert aliases to actual routes before the matcher sees them)

File: src/Symfony/Component/Routing/RouteCollection.php (hunk 4 above)
Comment (stof) [reply src/Symfony/Component/Routing/RouteCollection.php:378]: this is totally unnecessary, as the suggested phpdoc does not provide any additional info compared to the signature.

**Comment by fabpot** [comment]
We've just moved away from `master` as the main branch to use `5.x` instead. Unfortunately, I cannot reopen the PR and change the target branch to `5.x`. Can you open a new PR referencing this one to not loose the discussion? Thank you for your understanding and for your help.

**Comment by Lctrs** [comment]
https://github.com/symfony/symfony/pull/38464

**Pull request closed without merging**